telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala

```diff
@@ -0,0 +1,205 @@
+import com.twitter.finagle.zipkin.core.DeadlineSpanMap
+import com.twitter.finagle.Codec
+import java.io.ByteArrayOutputStream
+import java.io.ObjectOutputStream
+import java.io.IOException
+import org.apache.kafka.clients.producer.ProducerRecord
+import org.apache.kafka.clients.producer.Callback
+import org.apache.kafka.clients.producer.RecordMetadata
+import org.apache.thrift.TSerializer
+import com.twitter.finagle.zipkin.core.TracerCache
+
+import java.util.Map
+import java.util.HashMap
+
+import org.apache.kafka.clients.producer.ProducerConfig
+import com.twitter.logging.Logger
+
+class KafkaTelemeterInitializer extends TelemeterInitializer {
+  type Config = KafkaTelemeterConfig
+  override def configId = "io.buoyant.telemetry.KafkaTelemeterConfig"
+  def configClass = classOf[KafkaTelemeterConfig]
+}
+
+case class KafkaTelemeterConfig(brokerList: String, sampleRate: Float, numRetries: Int, metrics: Boolean, tracing: Boolean) extends TelemeterConfig {
+  private val log = Logger.get(getClass)
+  def mk(params: Stack.Params): KafkaTelemeter = {
+    log.info("Broker list is %s", brokerList)
+    log.info("Number of retries per request is %s", brokerList)
+    new KafkaTelemeter("zipkin", numRetries, sampleRate, brokerList, metrics, tracing)
+  }
+}
+
+//Going with hybrid approach. Let kafka handle retries. Batching will be handled outside as it is being done today.
+//We will leave linger.ms unchanged for this reason at 1 ms.
+//The reason for this is that the API has been constructed in such a way that a given trace is annotated across multiple
+//different calls rather than gathering up all its annotations and then sending it on its way.
+
+case class KafkaTelemeter(topic: String, numRetries: Int, sampleRate: Float, brokerList: String, metrics: Boolean, tracing: Boolean) extends Telemeter {
+  private val log = Logger.get(getClass)
+  log.info("Initializing kafka telemeter")
+  private var producer: KafkaProducer[Array[Byte], Array[Byte]] = null
+  private val METADATA_FETCH_TIMEOUT_CONFIG: Int = 3000
+  private val TIMEOUT_CONFIG: Int = 300
+  private val MAX_BLOCK_MS: Int = 1000
+  private val RETRY_BACKOFF_MS_CONFIG: Int = 10000
+  private val RECONNECT_BACKOFF_MS_CONFIG: Int = 10000
+  private val CACHE_BYTE_ARRAYS: Int = 100
+  //TODO need to do the tracerCache stuff to prevent duplicates in the config as well as allow support for multiple
+  //kafka clusters. TODO track all the different instances of tracers/dedup them.
+
+  def tracer: Tracer = {
+    val tracerImpl = KafkaRawZipkinTracer.tracerCache.putIfAbsent(
+      brokerList + topic,
+      new KafkaTracer(new KafkaRawZipkinTracer(brokerList, numRetries, topic), sampleRate)
+    )
+    KafkaRawZipkinTracer.tracerCache.get(brokerList + topic)
+  }
+
+  def cleanup(): Unit = {
+    log.info("Starting cleanup of the kafka telemeter")
+    //This is a stop gap arrangement. Give one minute for all remaining spans to be flushed.
+    Thread.sleep(1 * 60 * 1000)
```

Comment:
Let's extract the timeout constant and use the close(timeout, duration) instead.